[FSDP][state_dict] Return tensors instead of FlatParameters to avoid pickling errors

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94494
* __->__ #94637

After https://github.com/pytorch/pytorch/pull/88913, user-defined parameter states will be pickled. For a FlatParameter, this means `_local_shard` will also be pickled. Since state_dict and load_state_dict only require the tensor, returning the full FlatParameter does not give us any extra benefit. This PR changes the behavior to simply return a view of the FlatParameter.

Differential Revision: [D43205127](https://our.internmc.facebook.com/intern/diff/D43205127/)

fegin committed:
[FSDP][state_dict] Return tensors instead of FlatParameters to avoid pickling errors

After https://github.com/pytorch/pytorch/pull/88913, user-defined parameter states will be pickled. For a FlatParameter, this means `_local_shard` will also be pickled. Since state_dict and load_state_dict only require the tensor, returning the full FlatParameter does not give us any extra benefit. This PR changes the behavior to simply return a view of the FlatParameter.

Differential Revision: [D43205127](https://our.internmc.facebook.com/intern/diff/D43205127/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -393,8 +393,7 @@ def _local_post_state_dict_hook(
     if valid_data_size > 0:
-        if flat_param._shard_numel_padded > 0:
-            flat_param = flat_param.narrow(0, 0, valid_data_size)
+        flat_param = flat_param.view(valid_data_size)
```
nice, maybe add a comment saying this will make it return a tensor that can be properly serialized?

rohan-varma reviewed (approved):
thanks for the fix!

fegin committed:
Update on "[FSDP][state_dict] Return tensors instead of FlatParameters to avoid pickling errors"


After https://github.com/pytorch/pytorch/pull/88913, user-defined parameter states will be pickled. For a FlatParameter, this means `_local_shard` will also be pickled. Since state_dict and load_state_dict only require the tensor, returning the full FlatParameter does not give us any extra benefit. This PR changes the behavior to simply return a view of the FlatParameter.

Differential Revision: [D43205127](https://our.internmc.facebook.com/intern/diff/D43205127/)

[ghstack-poisoned]

fegin commented:
@pytorchbot  merge -f "The failing tests are not related."
